skiko/src/jvmMain/cpp/windows/directXRedrawer.cc

```diff
@@ -0,0 +1,343 @@
+        {
+            return 0;
+        }
+
+        d3dDevice->backendContext.fAdapter = hardwareAdapter;
+        d3dDevice->backendContext.fDevice = device;
+        d3dDevice->backendContext.fQueue = queue;
+
+        d3dDevice->device = device;
+        d3dDevice->queue = queue;
+
+        // Make the swapchain
+        HWND fWindow = (HWND)windowHandle;
+
+        gr_cp<IDXGIFactory4> swapChainFactory;
+        GR_D3D_CALL_ERRCHECK(CreateDXGIFactory2(0, IID_PPV_ARGS(&swapChainFactory)));
+
+        DXGI_SWAP_CHAIN_DESC1 swapChainDesc = {};
+        swapChainDesc.BufferCount = BuffersCount;
+        swapChainDesc.Format = DXGI_FORMAT_R8G8B8A8_UNORM;
+        swapChainDesc.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+        swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_FLIP_SEQUENTIAL;
+        swapChainDesc.SampleDesc.Count = 1;
+        swapChainDesc.SampleDesc.Quality = 0;
+        swapChainDesc.Scaling = DXGI_SCALING_NONE;
+
+        DXGI_SWAP_CHAIN_FULLSCREEN_DESC swapChainFSDesc = {};
+        swapChainFSDesc.Windowed = TRUE;
+
+        gr_cp<IDXGISwapChain1> swapChain;
+        GR_D3D_CALL_ERRCHECK(swapChainFactory->CreateSwapChainForHwnd(d3dDevice->queue.get(), fWindow, &swapChainDesc, &swapChainFSDesc, nullptr, &swapChain));
+        GR_D3D_CALL_ERRCHECK(swapChain->QueryInterface(IID_PPV_ARGS(&d3dDevice->swapChain)));
+        GR_D3D_CALL_ERRCHECK(d3dDevice->device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&d3dDevice->fence)));
+        d3dDevice->fenceEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+
+        return toJavaPointer(d3dDevice);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_resizeBuffers(
+        JNIEnv *env, jobject redrawer, jlong devicePtr, jint width, jint height)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+        if (!d3dDevice->isSizeEqualTo(width, height))
+        {
+            GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->ResizeBuffers(BuffersCount, width, height, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
+            d3dDevice->bufferWidth = width;
+            d3dDevice->bufferHeight = height;
+        }
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_finishFrame(
+        JNIEnv *env, jobject redrawer, jlong devicePtr, jlong contextPtr, jlong surfacePtr)
+    {
+        DirectXDevice *d3dDevice = fromJavaPointer<DirectXDevice*>(devicePtr);
+
+        SkSurface *surface = fromJavaPointer<SkSurface*>(surfacePtr);
+        surface->flushAndSubmit();
+        GrDirectContext *fContext = fromJavaPointer<GrDirectContext*>(contextPtr);
+        surface->flush(SkSurface::BackendSurfaceAccess::kPresent, GrFlushInfo());
+        fContext->flush({});
+        fContext->submit(true);
+
+        // 1 value in [Present(1, 0)] enables vblank wait so this is how vertical sync works in DirectX.
+        GR_D3D_CALL_ERRCHECK(d3dDevice->swapChain->Present(1, 0));
```

Comment:
We can support running without vsync.

For that we need to pass SkiaLayerProperties:
```
internal class Direct3DRedrawer(
    private val layer: HardwareLayer,
    private val properties: SkiaLayerProperties
) : Redrawer {
```

and pass bool parameter:
```
    external fun finishFrame(device: Long, context: Long, surface: Long) {
        finishFrame(device, context, surface, properties.isVsyncEnabled)
    }

    private external fun finishFrame(device: Long, context: Long, surface: Long, isVsyncEnabled: Boolean)
```

Reply:
Added.